### DISCUSSION TITLE
In FastAPI 0.71.0 and later, adding a middleware to the root app causes the system to swallow stack traces from sub-applications

### DISCUSSION THREAD

**Discussion by waweber** [fastapi/fastapi#8577] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

# main and sub-app
main_app = FastAPI()
sub_app = FastAPI()

# Add a view to sub-app that raises an error
@sub_app.get("/")
async def fail():
    raise ValueError("oops")

# mount the sub app at /test
# note: doing this after the @app.middleware does not change the problem
main_app.mount("/test", sub_app)

# no-op middleware on main app
@main_app.middleware("http")
async def middleware(request, handler):
    return await handler(request)

# Now run me like `uvicorn example:main_app --reload --port 8000`
# GET /test/
# Notice that the server just prints 500 Internal Server Error (what was the error?)
# Now comment out the middleware
# Notice that the server prints the stack trace now
```

### Description

When an unhandled exception occurs in a route handler, the server automatically responds with a 500 error response. It also prints the stack trace to the log, so the developer can tell what went wrong.

However, if you mount a sub-application AND add a middleware to the root application, exceptions raised in the sub-application will NOT print the stack trace, you will only see a "500 Internal Server Error" logged.

This behavior started in 0.71.0.

### Operating System

Linux

### FastAPI Version

Anything after 0.71.0

### Python Version

3.9.10

### Additional Context

I feel like this has something to do with Starlette's error handling middleware?

The answer might be "yeah don't set up your application that way" but if that's the case it would be helpful to warn about some middleware/subapplication quirks in the documentation.

**Comment by AlrasheedA** [comment]
Hi @waweber, faced the same issue and using [the suggestion](https://github.com/tiangolo/fastapi/issues/1802#issuecomment-790990222) from @haomingyin on issue #1802 worked.

**Comment by waza-ari** [comment]
I can confirm this behaviour, having the same problem in my environment.

**Comment by devamin** [comment]
This issue still persists with latest versions of `uvicorn==0.20.0` `fastapi==0.89.1` `starlette==0.22.0`

**Comment by exxy** [comment, marked as answer]
This issue appears to be resolved with latest versions of `uvicorn==0.24.0.post1` `fastapi==0.104.1` `starlette==0.27.0`

**Reply by devamin** [reply]
I confirm

**Reply by liniribeiro** [reply]
I tried to use this combination and the problem is still happening

**Comment by mskrip** [comment]
This issue is once again present with latest versions of `uvicorn==0.29.0` `fastapi==0.110.0` `starlette==0.36.3`. Can confirm that versions mentioned above in https://github.com/tiangolo/fastapi/discussions/8577#discussioncomment-7603730 fix the issue

**Reply by PiotrDabrowskey** [reply]
can confirm it still persists with
```
fastapi==0.110.2
starlette==0.37.2
uvicorn==0.29.0
```

**Reply by b4stien** [reply]
Same with the following versions. Has an issue been re-opened?

```
fastapi==0.111.0
starlette==0.37.2
uvicorn==0.30.1
```

**Reply by b4stien** [reply]
This is the last starlette/FastAPI versions that works (it works regardless of uvicorn):

```
fastapi==0.107.0
starlette==0.28.0
```

Starting with `fastapi==0.108` and `starlette==0.32.0.post1`, it breaks.

**Reply by b4stien** [reply]
[This is the PR that re-introduced this issue.](https://github.com/encode/starlette/pull/2194/files)

**Reply by DaanRademaker** [reply]
This issue is also present for me.
```
fastapi==0.109.2
starlette=0.36.3
```

Is there a new issue created for this? 

For custom logging middleware I need to catch the exceptions in the root app. 
I am using mounted apps in most of my applications. Am I misusing fastapi that this behavior is not supported anymore?

**Reply by benmoveai** [reply]
We're also seeing this issue on the following versions. this?

```
fastapi==0.111.0
starlette==0.37.2
uvicorn==0.26.0
```

Looks like [this](https://github.com/encode/starlette/issues/2625) is the issue that's tracking the fix

**Comment by Ilya-Q** [comment]
Seems to be happening again on the latest versions, i.e.:
```
fastapi==0.116.1
starlette==0.47.3
uvicorn==0.35.0
```
